File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/contacts/ContactDetailScreen.kt
package app.getvela.wallet.feature.contacts

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import app.getvela.wallet.core.designsystem.tokens.VelaIconSize
import app.getvela.wallet.core.designsystem.tokens.VelaRadius
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import app.getvela.wallet.core.designsystem.components.VelaIcons
import app.getvela.wallet.core.designsystem.theme.VelaTheme
import app.getvela.wallet.core.designsystem.tokens.VelaFontFamily
import app.getvela.wallet.core.designsystem.tokens.VelaFontWeight
import app.getvela.wallet.core.designsystem.tokens.VelaMonoFontFamily
import app.getvela.wallet.core.designsystem.tokens.VelaSizing
import app.getvela.wallet.core.designsystem.tokens.VelaSpacing
import app.getvela.wallet.core.designsystem.tokens.VelaTextSize
import app.getvela.wallet.feature.contacts.components.AddressBlock
import app.getvela.wallet.feature.contacts.components.ContactsMetrics
import app.getvela.wallet.feature.contacts.components.ContactsNavHeader
import app.getvela.wallet.feature.contacts.components.DeleteConfirmSheet
import app.getvela.wallet.feature.contacts.components.DestructiveTextButton
import app.getvela.wallet.feature.contacts.components.EmptyStateCta
import app.getvela.wallet.feature.contacts.components.GroupChips
import app.getvela.wallet.feature.contacts.components.Hairline
import app.getvela.wallet.feature.contacts.components.glyphFor
import app.getvela.wallet.feature.wallet.components.ActionButtonItem
import app.getvela.wallet.feature.wallet.components.ActionButtonRow
import app.getvela.wallet.feature.wallet.components.ActivityRow
import app.getvela.wallet.feature.wallet.components.IdenticonAvatar
import app.getvela.wallet.feature.wallet.components.SectionHeader

/**
 * Contact detail (mocks C2 / C2s): back + pencil header, hero identicon, name,
 * short address, group chips, the reused three-card action dock, the mono
 * 地址 block, 最近往来 over spec-015 ActivityRows, and the centred destructive
 * 删除联系人 that raises the confirmation sheet.
 */
@Composable
fun ContactDetailScreen(
    model: ContactsHomeModel,
    modifier: Modifier = Modifier,
    actions: ContactsActions = ContactsActions(),
) {
    val detail = model.detail ?: return
    val colors = VelaTheme.colors
    Box(
        modifier = modifier
            .fillMaxSize()
            .background(colors.bgBase),
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .statusBarsPadding()
                .verticalScroll(rememberScrollState())
                .navigationBarsPadding()
                .padding(horizontal = VelaSizing.screenPaddingX),
        ) {
            Spacer(modifier = Modifier.height(VelaSpacing.md))
            ContactsNavHeader(
                backContentDescription = model.backLabel,
                trailingIcon = VelaIcons.Pencil,
                trailingContentDescription = detail.editLabel,
                onBack = { actions.onAction("contacts.back") },
                onTrailing = { actions.onAction("contacts.edit") },
            )

            Spacer(modifier = Modifier.height(VelaSpacing.xl3))
            Column(
                modifier = Modifier.align(Alignment.CenterHorizontally),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                IdenticonAvatar(
                    seed = detail.contact.addressFull,
                    size = ContactsMetrics.heroAvatar,
                    contentDescription = detail.contact.name,
                )
                Spacer(modifier = Modifier.height(VelaSpacing.xl))
                Text(
                    text = detail.contact.name,
                    color = colors.fgBase,
                    fontFamily = VelaFontFamily,
                    fontWeight = VelaFontWeight.bold,
                    fontSize = VelaTextSize.xl3,
                    maxLines = 1,
                )
                Spacer(modifier = Modifier.height(VelaSpacing.md))
                Text(
                    text = detail.contact.addressDisplay,
                    color = colors.fgMuted,
                    fontFamily = VelaMonoFontFamily,
                    fontWeight = VelaFontWeight.regular,
                    fontSize = VelaTextSize.base,
                    maxLines = 1,
                )
                detail.inspection?.let { inspection ->
                    // Spec 045 US7: the core's word on this address — contract or
                    // wallet, and whether the wallet has paid it before.
                    Spacer(modifier = Modifier.height(VelaSpacing.md))
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        inspection.tag?.let { tag ->
                            Text(
                                text = tag,
                                color = colors.fgMuted,
                                fontFamily = VelaFontFamily,
                                fontSize = VelaTextSize.xs,
                                modifier = Modifier
                                    .background(colors.bgRaised, RoundedCornerShape(VelaRadius.full))
                                    .padding(horizontal = VelaSpacing.md, vertical = VelaSpacing.xs)
                                    .semantics { contentDescription = "contact-inspection-tag" },
                            )
                        }
                        inspection.firstTime?.let { first ->
                            if (inspection.tag != null) Spacer(modifier = Modifier.width(VelaSpacing.sm))
                            Text(
                                text = first,
                                color = colors.fgSubtle,
                                fontFamily = VelaFontFamily,
                                fontSize = VelaTextSize.xs,
                            )
                        }
                    }
                }
                detail.favourite?.let { favourite ->
                    // Spec 045 US5: the favourite star — 018's section word as its label.
                    Spacer(modifier = Modifier.height(VelaSpacing.md))
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier
                            .clip(RoundedCornerShape(VelaRadius.full))
                            .clickable { actions.onAction("contacts.favourite") }
                            .padding(horizontal = VelaSpacing.md, vertical = VelaSpacing.xs)
                            .semantics { contentDescription = if (favourite.on) "favourite-on" else "favourite-off" },
                    ) {
                        Icon(
                            imageVector = VelaIcons.Star,
                            contentDescription = null,
                            tint = if (favourite.on) colors.accentBase else colors.fgSubtle,
                            modifier = Modifier.size(VelaIconSize.sm),
                        )
                        Spacer(modifier = Modifier.width(VelaSpacing.xs))
                        Text(
                            text = favourite.label,
                            color = if (favourite.on) colors.accentBase else colors.fgMuted,
                            fontFamily = VelaFontFamily,
                            fontSize = VelaTextSize.sm,
                        )
                    }
                }
                Spacer(modifier = Modifier.height(VelaSpacing.lg))
                GroupChips(
                    model = detail.chips,
                    onGroup = { actions.onAction("contacts.sectionGroups") },
                    onAdd = { actions.onAction("contacts.moveGroup") },
                )
            }

            Spacer(modifier = Modifier.height(VelaSpacing.xl3))
            ActionButtonRow(
                items = detail.actions.map { action ->
                    ActionButtonItem(
                        icon = glyphFor(action.icon),
                        label = action.label,
                        onClick = { actions.onAction("contacts.action." + action.icon.name) },
                    )
                },
            )

            Spacer(modifier = Modifier.height(VelaSpacing.xl3))
            Hairline()
            Spacer(modifier = Modifier.height(VelaSpacing.xl3))
            AddressBlock(
                model = detail.address,
                onCopy = { actions.onAction("contacts.copyAddress") },
            )

            Spacer(modifier = Modifier.height(VelaSpacing.xl3))
            SectionHeader(
                title = detail.activity.title,
                action = detail.activity.action,
                onAction = { actions.onAction("history.filterAll") },
            )
            if (detail.activity.rows.isEmpty()) {
                detail.activity.empty?.let {
                    EmptyStateCta(model = it, icon = VelaIcons.Inbox)
                }
            } else {
                detail.activity.rows.forEach { row -> ActivityRow(model = row) }
            }

            Spacer(modifier = Modifier.height(VelaSpacing.xl4))
            DestructiveTextButton(
                label = detail.deleteLabel,
                onClick = { actions.onAction("contacts.deleteContact") },
            )
            Spacer(modifier = Modifier.height(VelaSpacing.xl4))
        }
    }

    model.deleteConfirm?.let { confirm ->
        DeleteConfirmSheet(
            model = confirm,
            onDismiss = actions.onDismissMenu,
            onConfirm = { actions.onAction("contacts.delete") },
        )
    }
}

File: app-ios/VelaWallet/VelaWallet/Features/Contacts/ContactDetailScreen.swift
//
//  ContactDetailScreen.swift
//  VelaWallet
//
//  Mobile contact detail (spec 018 FR-002, mocks C2 / C2s): back chevron +
//  edit pencil, identicon hero, name, short address, group chips, the three
//  action cards, the full mono address block, 最近往来 over spec-015
//  ActivityRow instances, and the centered destructive 删除联系人 which
//  raises the confirm sheet (C2s).
//
//  Reused from spec 015: IdenticonAvatar, ActionButtonRow, WalletSectionHeader,
//  ActivityRowView, WalletEmptyState (the no-activity edge case).
//

import SwiftUI

struct ContactDetailScreen: View {
    @Environment(\.theme) private var theme

    let model: ContactDetailModel
    var onBack: () -> Void = {}
    /// The pencil. Opens the edit form, and it shipped doing nothing.
    var onEdit: () -> Void = {}
    /// The star. Every client's `ToggleFavorite`.
    var onFavourite: () -> Void = {}
    var onDelete: () -> Void = {}
    /// The live form's fields and CTA. Absent in the gallery, where the sheet
    /// is a picture.
    var formName: Binding<String>?
    var formAddress: Binding<String>?
    var onSaveForm: () -> Void = {}
    var onCancelForm: () -> Void = {}
    /// + 分组 — open the group picker, tick a row, commit the set.
    var onOpenGroups: () -> Void = {}
    var onToggleGroup: (String) -> Void = { _ in }
    var onSaveGroups: () -> Void = {}
    var onCancelGroups: () -> Void = {}
    /// 转账 — a send already addressed to this person.
    var onSendTo: (() -> Void)?
    /// 收款 and 二维码, which are about the wallet's OWN address.
    var onReceive: (() -> Void)?
    var onShowQr: (() -> Void)?

    /// Whether the delete confirmation is up. The form's and the picker's
    /// presence are the CORE's answers and need no flags of their own.
    @State private var confirmShown = false

    /// What the one sheet is showing, in order of precedence: a form or a
    /// picker is open because somebody asked for it just now.
    private enum Presented {
        case confirm(ActionMenuModel)
        case form(ContactFormModel)
        case groups(MultiPickModel)
    }

    private var presented: Presented? {
        if let form = model.form { return .form(form) }
        if let pick = model.groupPick { return .groups(pick) }
        if confirmShown, let sheet = model.sheet { return .confirm(sheet) }
        return nil
    }

    private func dismissSheet() {
        if model.form != nil { onCancelForm() }
        if model.groupPick != nil { onCancelGroups() }
        confirmShown = false
    }

    var body: some View {
        VStack(spacing: Tokens.Space.s0) {
            navBar
            ScrollView {
                VStack(alignment: .leading, spacing: Tokens.Space.s0) {
                    hero
                    // Three cards that did nothing until 057. 转账 opens a
                    // send already addressed to this person; 收款 and 二维码
                    // are about the WALLET's own address, so they go where that
                    // lives rather than pretending to be about the contact.
                    ActionButtonRow(items: [
                        ActionCardItem(icon: .arrowUpRight, label: model.actions.send,
                                       action: onSendTo),
                        ActionCardItem(icon: .arrowDownLeft, label: model.actions.receive,
                                       action: onReceive),
                        ActionCardItem(icon: .qrCode, label: model.actions.qr,
                                       action: onShowQr),
                    ])
                    .padding(.top, Tokens.Space.s24)

                    hairline.padding(.top, Tokens.Space.s24)

                    AddressBlock(
                        label: model.addressLabel,
                        lines: model.addressLines,
                        copyLabel: model.copyLabel,
                        copiedLabel: model.copiedLabel
                    )
                    .padding(.top, Tokens.Space.s24)

                    hairline.padding(.top, Tokens.Space.s24)

                    WalletSectionHeader(title: model.activityTitle, action: model.activityAction)
                        .padding(.top, Tokens.Space.s24)
                    activity

                    deleteAction
                        .padding(.top, Tokens.Space.s32)
                }
                .padding(.horizontal, Tokens.Layout.screenPaddingX)
                .padding(.bottom, Tokens.Space.s32)
            }
        }
        .background(theme.bgBase.ignoresSafeArea())
        .walletTextScale(model.textScale)
        // ONE sheet, whose CONTENT changes. Three `.sheet` modifiers on one
        // view is what stopped the contacts list responding to taps at all
        // (the same defect, on the screen before this one).
        .sheet(isPresented: Binding(
            get: { presented != nil },
            set: { if !$0 { dismissSheet() } }
        )) {
            switch presented {
            case .confirm(let menu):
                ActionMenuSheet(
                    model: menu,
                    // The one item this sheet has is the destructive one: it is
                    // the second confirmation on 删除联系人, and until spec 054
                    // it went nowhere — the button raised a sheet the live
                    // builder left empty, which is a confirmation dialog for a
                    // deletion that never happened.
                    onItem: { item in
                        confirmShown = false
                        if item.destructive { onDelete() }
                    },
                    onCancel: { confirmShown = false }
                )
                .walletTextScale(model.textScale)
            case .form(let form):
                ContactFormSheet(
                    model: form,
                    nameText: formName,
                    addressText: formAddress,
                    onSave: onSaveForm,
                    onCancel: { onCancelForm() }
                )
                .walletTextScale(model.textScale)
            case .groups(let pick):
                MultiPickSheet(
                    model: pick,
                    onToggle: onToggleGroup,
                    onSave: onSaveGroups,
                    onCancel: { onCancelGroups() }
                )
                .walletTextScale(model.textScale)
            case nil:
                EmptyView()
            }
        }
        // C2s is the state that opens WITH the confirm up; every other one has
        // it behind 删除联系人 and starts closed.
        .onAppear { confirmShown = model.state == .c2s }
    }

    private func inspectionTag(_ text: String) -> some View {
        Text(verbatim: text)
            .typeRole(Typography.rowSub.scaled(model.textScale))
            .foregroundStyle(theme.fgMuted)
            .padding(.horizontal, Tokens.Space.s8)
            .padding(.vertical, Tokens.Space.s2)
            .background(Capsule().fill(theme.bgRaised))
    }

    // MARK: - Chrome

    private var navBar: some View {
        HStack(spacing: Tokens.Space.s12) {
            Button(action: onBack) {
                LucideIcon(.chevronLeft, size: LucideIconSize.action)
                    .foregroundStyle(theme.fgBase)
                    .frame(width: Tokens.Layout.hitTarget, height: Tokens.Layout.hitTarget)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(model.backLabel)
            Spacer(minLength: Tokens.Space.s12)
            if let favourite = model.favourite {
                Button(action: onFavourite) {
                    LucideIcon(favourite.on ? .starSolid : .star, size: LucideIconSize.menuRow)
                        .foregroundStyle(favourite.on ? theme.accentBase : theme.fgMuted)
                        .frame(width: Tokens.Layout.hitTarget, height: Tokens.Layout.hitTarget)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityLabel(favourite.label)
                .accessibilityAddTraits(favourite.on ? [.isButton, .isSelected] : .isButton)
            }
            Button(action: onEdit) {
                LucideIcon(.pencil, size: LucideIconSize.menuRow)
                    .foregroundStyle(theme.fgBase)
                    .frame(width: Tokens.Layout.hitTarget, height: Tokens.Layout.hitTarget)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(model.editLabel)
        }
        .padding(.horizontal, Tokens.Space.s12)
    }

    private var hero: some View {
        VStack(spacing: Tokens.Space.s0) {
            IdenticonAvatar(seed: model.contact.addressFull, size: ContactsGeometry.detailAvatar,
                             name: model.contact.name)
            Text(verbatim: model.contact.name)
                .typeRole(Typography.title.scaled(model.textScale))
                .foregroundStyle(theme.fgBase)
                .multilineTextAlignment(.center)
                .padding(.top, Tokens.Space.s16)
            Text(verbatim: model.contact.addressDisplay)
                .monoRole(Typography.monoAddressDetail.scaled(model.textScale))
                .foregroundStyle(theme.fgMuted)
                .padding(.top, Tokens.Space.s8)
            // What the core found out about this address when the page opened:
            // whether it is a contract wallet, and whether this person has ever
            // been paid before. Two neutral tags — never a warning, because
            // neither fact is one.
            if let inspection = model.inspection,
               inspection.tag != nil || inspection.firstTime != nil {
                HStack(spacing: Tokens.Space.s8) {
                    if let tag = inspection.tag { inspectionTag(tag) }
                    if let first = inspection.firstTime { inspectionTag(first) }
                }
                .padding(.top, Tokens.Space.s8)
            }
            GroupChips(chips: model.chips, addLabel: model.addChip, onAdd: onOpenGroups)
                .padding(.top, Tokens.Space.s12)
        }
        .frame(maxWidth: .infinity)
        .padding(.top, Tokens.Space.s16)
    }

    @ViewBuilder private var activity: some View {
        if model.activity.isEmpty, let empty = model.activityEmpty {
            WalletEmptyState(icon: .inbox, model: empty)
        } else {
            VStack(spacing: Tokens.Space.s0) {
                ForEach(Array(model.activity.enumerated()), id: \.element.id) { index, row in
                    if index > 0 {
                        Rectangle()
                            .fill(theme.borderBase)
                            .frame(height: Tokens.BorderWidth.hairline)
                            .padding(.leading, WalletGeometry.rowDividerInset)
                    }
                    ActivityRowView(model: row)
                }
            }
        }
    }

    private var deleteAction: some View {
        Button {
            confirmShown = true
        } label: {
            Text(verbatim: model.deleteLabel)
                .typeRole(Typography.button.scaled(model.textScale))
                .foregroundStyle(theme.errorBase)
                .frame(maxWidth: .infinity)
                .frame(minHeight: Tokens.Control.md)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private var hairline: some View {
        Rectangle()
            .fill(theme.borderBase)
            .frame(height: Tokens.BorderWidth.hairline)
    }
}

#Preview("C2 detail dark") {
    ContactDetailScreen(model: ContactsFixtures.buildMobileState(.c2, loc: ContactsPreviewData.loc).detail!)
        .themed(.dark)
        .environment(\.identiconProvider, .previewSafe)
        .environment(\.lucideIconProvider, .previewSafe)
}

#Preview("C2 detail light") {
    ContactDetailScreen(model: ContactsFixtures.buildMobileState(.c2, loc: ContactsPreviewData.loc).detail!)
        .themed(.light)
        .environment(\.identiconProvider, .previewSafe)
        .environment(\.lucideIconProvider, .previewSafe)
}
